espnet2/asr/discrete_asr_espnet_model.py

```diff
@@ -198,62 +204,110 @@ def forward(
         loss, stats, weight = force_gatherable((loss, stats, batch_size), loss.device)
         return loss, stats, weight
 
-    def encode(
-        self, src_text: torch.Tensor, src_text_lengths: torch.Tensor
+    def encode(self, 
+        src_text: torch.Tensor, 
+        src_text_lengths: torch.Tensor,
+        hubert_feats: Optional[torch.Tensor] = None,  
+        hubert_feats_lengths: Optional[torch.Tensor] = None
     ) -> Tuple[torch.Tensor, torch.Tensor]:
         """Frontend + Encoder. Note that this method is used by mt_inference.py
 
         Args:
             src_text: (Batch, Length, ...)
             src_text_lengths: (Batch, )
+            hubert: (Batch, Length, ...)  # Hubert特征
+            hubert_lengths: (Batch, )  # Hubert特征长度
         """
+        
         with autocast(False):
             # 1. Extract feats
             feats, feats_lengths = self._extract_feats(src_text, src_text_lengths)
 
+            # 提供了Hubert特征，合并或处理它们
+            if hubert_feats is not None:
+                # 确保Hubert特征符合预期形状和长度,将它们直接与前端特征拼接
+                feats = torch.cat((feats, hubert_feats), dim=-1)
```

Comment:
??? are you sure that the current code is runnable? I did not see where the `hubert_feats` is prepared.

Reply:
<img width="773" alt="image" src="https://github.com/leonardo-xing/espnet/assets/58717132/07d095e1-8d8c-47fa-b70a-7a272102ee36">
The hubert_feature is passed in from the encoder, which gets the data from the opts variable. This one is my understanding.